cmd/download.go

```diff
@@ -7,150 +7,252 @@ package cmd
 import (
 	"fmt"
 	"os"
-	"runtime"
+	"path"
 	"strings"
 	"sync"
 
+	"github.com/hashicorp/go-getter"
 	"github.com/sirupsen/logrus"
+)
 
-	getter "github.com/hashicorp/go-getter"
+var (
+	fallbackHttpsRepoPrefix = "git::https://github.com/sighupio/fury-kubernetes"
+	fallbackSshRepoPrefix   = "git@github.com:sighupio/fury-kubernetes"
+	httpsRepoPrefix         = "git::https://github.com/sighupio/kubernetes-fury"
+	sshRepoPrefix           = "git@github.com:sighupio/kubernetes-fury"
 )
 
-var parallel bool
-var https bool
-var prefix string
+type DownloadOpts struct {
+	Parallel bool
+	Https    bool
+	Fallback bool
+}
 
-func download(packages []Package) error {
+//PackageURL is the representation of the fields needed to elaborate a url
+type PackageURL struct {
+	Prefix        string
+	Blocks        []string
+	Kind          string
+	Version       string
+	Registry      bool
+	CloudProvider ProviderOptSpec
+	KindSpec      ProviderKind
+}
 
-	// Preparing all the necessary data for a worker pool
-	var wg sync.WaitGroup
-	var numberOfWorkers int
-	if parallel {
-		numberOfWorkers = runtime.NumCPU() + 1
+// newUrl initialize the PackageURL struct
+func newPackageURL(prefix string, blocks []string, kind, version string, registry bool, cloud ProviderOptSpec, kindSpec ProviderKind) *PackageURL {
+	return &PackageURL{
+		Prefix:        prefix,
+		Registry:      registry,
+		Blocks:        blocks,
+		Kind:          kind,
+		Version:       version,
+		CloudProvider: cloud,
+		KindSpec:      kindSpec,
+	}
+}
+
+//getConsumableURL returns an url that can be used for download
+func (n *PackageURL) getConsumableURL() string {
+
+	if !n.Registry {
+		return n.getURLFromCompanyRepos()
+	}
+
+	return fmt.Sprintf("%s/%s%s?ref=%s", n.KindSpec.pickCloudProviderURL(n.CloudProvider), n.Blocks[0], ".git", n.Version)
+
+}
+
+func (n *PackageURL) getURLFromCompanyRepos() string {
+	if len(n.Blocks) == 0 {
+		// todo should return error?
+		return ""
+	}
+
+	dG := ""
+
+	if strings.HasPrefix(n.Prefix, "git::https") {
+		dG = ".git"
+	}
+
+	if len(n.Blocks) == 1 {
+		return fmt.Sprintf("%s-%s%s//%s?ref=%s", n.Prefix, n.Blocks[0], dG, n.Kind, n.Version)
+	}
+	// always len(n.Blocks) >= 2 {
```

Comment:
still needed?

Reply:
I think not but it's a legacy code so I would leave it for now.